src/sage/schemes/curves/plane_curve_arrangement.py

```diff
@@ -0,0 +1,1342 @@
+        return self(x)
+
+    @cached_method
+    def ngens(self):
+        """
+        Return the number of variables, i.e. 2 or 3, kept for completness.
+
+        OUTPUT:
+
+        An integer, 2 or 3, depending if the arrangement is projective or affine.
+
+        EXAMPLES::
+
+            sage: L.<x, y> = AffinePlaneCurveArrangements(QQ)
+            sage: L.ngens()
+            2
+            sage: L.<x, y, z> = ProjectivePlaneCurveArrangements(QQ)
+            sage: L.ngens()
+            3
+        """
+        return len(self.variable_names())
+
+    @cached_method
+    def gens(self):
+        """
+        Return the coordinates.
+
+        OUTPUT:
+
+        A tuple of linear expressions, one for each linear variable.
+
+        EXAMPLES::
+
+            sage: L = AffinePlaneCurveArrangements(QQ, ('x', 'y'))
+            sage: L.gens()
+            (x, y)
+            sage: L = ProjectivePlaneCurveArrangements(QQ, ('x', 'y', 'z'))
+            sage: L.gens()
+            (x, y, z)
+        """
+        return self.ambient_space().gens()
+
+    def gen(self, i):
+        """
+        Return the `i`-th coordinate.
+
+        INPUT:
+
+        - ``i`` -- integer
+
+        OUTPUT:
+
+        A variable.
+
+        EXAMPLES::
+
+            sage: L.<x, y> = AffinePlaneCurveArrangements(QQ)
+            sage: L.gen(1)
+            y
+            sage: L.<x, y, z> = ProjectivePlaneCurveArrangements(QQ)
+            sage: L.gen(2)
+            z
+        """
+        return self.gens()[i]
+
+
+class AffinePlaneCurveArrangements(PlaneCurveArrangements):
+    """
+    Affine curve arrangements.
+
+    INPUT:
+
+    - ``base_ring`` -- ring; the base ring
+
+    - ``names`` -- tuple of strings; the variable names
+
+    EXAMPLES::
+
+        sage: H.<x, y> = AffinePlaneCurveArrangements(QQ)
+        sage: H(x, y^2, x-1, y-1)
+        Arrangement (x, y^2, x - 1, y - 1) in Affine Space
+        of dimension 2 over Rational Field
+    """
+    Element = AffinePlaneCurveArrangementElement
+
+    def __init__(self, base_ring, names=tuple()):
+        """
+        Initialize ``self``.
+
+        TESTS::
+
+            sage: H.<x, y> = AffinePlaneCurveArrangements(QQ)
+            sage: TestSuite(H).run()
+        """
+        if base_ring not in _Fields:
+            raise ValueError('base ring must be a field')
+        super().__init__(base_ring, names=names)
+
+
+class ProjectivePlaneCurveArrangements(PlaneCurveArrangements):
+    """
+    Projective curve arrangements.
+
+    INPUT:
+
+    - ``base_ring`` -- ring; the base ring
+
+    - ``names`` -- tuple of strings; the variable names
+
+    EXAMPLES::
+
+        sage: H.<x, y, z> = ProjectivePlaneCurveArrangements(QQ)
+        sage: H(x, y^2, x-z, y-z)
+        Arrangement (x, y^2, x - z, y - z) in Projective Space
+        of dimension 2 over Rational Field
+    """
+    Element = ProjectivePlaneCurveArrangementElement
+
+    def __init__(self, base_ring, names=tuple()):
```

Comment:
Same comment as for affine. (Of course, the tests here should be kept and moved up there.)